gaugan/callbacks.py

```diff
@@ -31,32 +34,38 @@ def infer(self):
         )
         return self.model.predict([latent_vector, self.val_images[2]])
 
+    def log_to_tables(self, epoch, semantic_input, ground_truth, generated_images):
+        wandb_table = wandb.Table(columns=self.columns)
+        for i in range(4):
+            wandb_table.add_data(
+                wandb.Image(semantic_input[i]),
+                wandb.Image(ground_truth[i]),
+                wandb.Image(generated_images[i]),
+            )
+        wandb.log({f"GANMonitor Epoch {epoch}": wandb_table})
+
     def on_epoch_end(self, epoch, logs=None):
         if epoch == 0 or (epoch + 1) % self.epoch_interval == 0:
             generated_images = self.infer()
 
-            grid_row = min(generated_images.shape[0], 3)
-            fig, axarr = plt.subplots(4, 3, figsize=(18, grid_row * 6))
+            semantic_input = (self.val_images[0] + 1) / 2
+            ground_truth = (self.val_images[1] + 1) / 2
+            generated_images = (generated_images + 1) / 2
+
+            if self.use_wandb:
+                self.log_to_tables(
+                    epoch + 1, semantic_input, ground_truth, generated_images
+                )
 
             for i in range(4):
-                for j in range(3):
-                    # ax = axarr if grid_row == 1 else axarr[row]
-                    if j == 0:
-                        axarr[i, j].imshow((self.val_images[0][i] + 1) / 2)
-                        axarr[i, j].axis("off")
-                        axarr[i, j].set_title("Mask", fontsize=20)
-                    if j == 1:
-                        axarr[i, j].imshow((self.val_images[1][i] + 1) / 2)
-                        axarr[i, j].axis("off")
-                        axarr[i, j].set_title("Reference Image", fontsize=20)
-                    if j == 2:
-                        axarr[i, j].imshow((generated_images[i] + 1) / 2)
-                        axarr[i, j].axis("off")
-                        axarr[i, j].set_title("Generated Image", fontsize=20)
-
-            if (self.plot_save_dir is None) and (not self.use_wandb):
-                plt.show()
-            elif self.use_wandb:
-                wandb.log({f"validation_images_{epoch}": fig})
-            elif self.plot_save_dir:
-                fig.savefig(os.path.join(self.plot_save_dir, f"{epoch}.png"))
+                fig = plot_results(
+                    [semantic_input[i], ground_truth[i], generated_images[i]],
+                    self.columns,
+                    figure_size=(18, 18),
+                )
+                if (self.plot_save_dir is None) and (not self.use_wandb):
+                    plt.show()
+                elif self.use_wandb:
+                    wandb.log({f"validation_images_{epoch + 1}_{i}": fig})
```

Comment:
No need to use this if we are using `wandb.Table` I suppose?

Reply:
Done